### ISSUE TITLE
Crash with micro text editor after editing and saving files

### ISSUE THREAD

**Issue by ThGrSoRu** [sxyazi/yazi#456]
### What system are you running Yazi on?

Linux Wayland

### What terminal are you running Yazi in?

foot version: 1.16.2

### Yazi version

yazi 0.1.5 (82bab0f 2023-12-26)

### Did you try the latest main branch to see if the problem has already been fixed?

Tried, but the problem is still present

### Describe the bug

Opening a file with [micro](https://github.com/zyedidia/micro), editing a single character, saving and closing the file leads to:
```
Backtrace (most recent call first):

The application panicked (crashed).
  panic in a function that cannot unwind
in library/core/src/panicking.rs, line 136
thread: tokio-runtime-worker
```
This only seems to happen when files are at least 2 folders away from yazi's starting directory.

### Expected Behavior

Yazi should not crash.

### To Reproduce

See above.

**Comment by sxyazi** [comment]
I can't reproduce it on my side. Could you record a gif and let me see your steps?

**Comment by Brixy** [comment]
## System

Linux Wayland

## Terminal

foot 1.16.2

## Yazi version

yazi 0.1.5 (82bab0f 2023-12-26)

## Describe the bug

Not quite sure whether this is related; but the error message is **identical:**

As soon as I hover any file with a `json` or `lock` extension, yazi crashes with the error message above. So far I haven't found other extensions affected.

Thanks for your latest huge commit! Impressive.

**Comment by sxyazi** [comment]
I tried running the latest code on my macOS and Linux VM, but unfortunately, I still couldn't reproduce it.

When hovering over the `json` and `lock` files, they work fine, and I can edit them and preview the changes normally.

There's no information in the error message that could help pinpoint the issue. Could you please run Yazi in debug mode and set `RUST_BACKTRACE=1`, i.e. do a 

```bash
export RUST_BACKTRACE=1
cargo run
```

In Yazi's source code directory.

**Comment by Brixy** [comment]
Thank you. The yazi binary compiled with `export RUST_BACKTRACE=1` does not crash any more. Yet, there is no preview for these file types (`json`, `lock`).

`cargo run` echoes quite a number of warnings. Would posting them help you?

**Comment by sxyazi** [comment]
So weird. I want to know how you built Yazi before and if there are any configuration files, please paste them here.

> cargo run echoes quite a number of warnings. Would posting them help you?

That's unrelated; those are part of the plugin system that's still incomplete. I left them there to remind me to finish it. 🤣

**Comment by Brixy** [comment]
Usually I compile yazi with

```
cargo build --release
```

Error happens without any user config files, too.

But this is my current `yazi.toml`:

<details>

```
[manager]
layout         = [ 0, 4, 3 ]
sort_by        = "natural"
sort_sensitive = false
sort_reverse   = false
sort_dir_first = true
linemode       = "none"
show_hidden    = false
show_symlink   = true

[preview]
tab_size   = 2
max_width  = 600
max_height = 900
cache_dir  = ""

[opener]
open = [
	{ exec = 'xdg-open "$@"', orphan = true, desc = "open", for = "unix" },
]
editTerm = [
	{ exec = '$TERMINAL $EDITOR "$@"', orphan = true, desc = "edit in new terminal", for = "unix" },
]
edit = [
	{ exec = '$EDITOR "$@"', block = true, desc = "edit", for = "unix" },
]

[open]
rules = [
	{ name = "*", use = [ "open", "editTerm", "edit" ] },
]

[tasks]
micro_workers = 5
macro_workers = 10
bizarre_retry = 5

[plugins]
preload = []

[log]
enabled = false
```

</details>

Thank you so much for investigating!

**Comment by sxyazi** [comment]
I think I've pinpointed the issue, which might be a bug in mlua. I'm not sure how to handle it yet, so I've decided to temporarily disable it.

I made a PR to fix it, https://github.com/sxyazi/yazi/pull/459, please give it a try and let me know if it works for you.

**Comment by manfred3000** [comment]
@sxyazi The latest PR does fix crashing on renaming/editing files, but `yazi` still crashes when hovering over certain file types - but I haven't figured out which types AND this behavior is not consistent across different HW (always Arch Linux with `wayland`)
Thank you for your time and work!

**Comment by ThGrSoRu** [comment]
Same here. The PR fixes editing files with micro, but scrolling over text files with a preview causes the same crash.

**Comment by sxyazi** [comment]
Hi, I pushed a new commit. Please try it out again to see if it works. Hope it's effective :)

**Comment by Brixy** [comment]
> Hi, I pushed a new commit in that PR. Please try it out again to see if it works. Hope it's effective :)

Thanks a lot!

a005f20 solved this issue on my side.

**Comment by manfred3000** [comment]
> Hi, I pushed a new commit in that PR. Please try it out again to see if it works. Hope it's effective :)

`yazi` built with that commit doesn't "show" any files here at all (only previews), unfortunately...

**Comment by ThGrSoRu** [comment]
While that fixes the crash, some previews of text files, specifically .json files, are missing.

**Comment by sxyazi** [comment]
Okay, let's merge that PR first, and I'll keep this issue open to track any other potential issues.

**Comment by sxyazi** [comment]
As for `.json` preview, it's a breaking change - prior to Yazi v0.2.0, it would call `jq` first, and if that call failed (e.g., if the user hadn't installed it), it would fallback to Yazi's built-in code highlighting.

However, this behavior is inefficient for users who never installed `jq` but always use the built-in highlighting - it constantly creates processes to attempt whether `jq` is installed.

In Yazi v0.2.0, this fallback has been removed. When users don't really need `jq`, they can directly specify Yazi's built-in highlighting for the `.json` preview:

```toml
# yazi.toml
[plugin]
previewers = [
	# ...
	{ mime = "application/json", exec = "code.lua" },
]
```

Try to add this line to the `previewers`. For now, please [copy the entire `previewers`](https://github.com/sxyazi/yazi/blob/2d3512e965f4bd46d7fd8310424138b0c6cbe103/yazi-config/preset/yazi.toml#L89-L114). I will add new options before the release of Yazi v0.2.0 to allow users to selectively override certain previewers.

**Comment by sxyazi** [comment]
> > Hi, I pushed a new commit in that PR. Please try it out again to see if it works. Hope it's effective :)
> 
> `yazi` built with that commit doesn't "show" any files here at all (only previews), unfortunately... : (
> 
> EDIT: I used: `linemode = "mtime"` With: `linemode = "none"` files "reappeared"...

Should be fixed in https://github.com/sxyazi/yazi/commit/bfa5e0af0c5493d1420d68b3346be808177f496c, thanks for the report!

**Comment by manfred3000** [comment]
> > EDIT: I used: `linemode = "mtime"` With: `linemode = "none"` files "reappeared"...
> 
> Should be fixed in [bfa5e0a](https://github.com/sxyazi/yazi/commit/bfa5e0af0c5493d1420d68b3346be808177f496c), thanks for the report!

Yep, thanks a lot!

**Comment by sxyazi** [comment]
I think this issue has been resolved, so let me close it now.

BTW, in order to reduce the impact of v0.2.0 on user behavior, I've decided to reintroduce the fallback. 🤣

However, for users who never use `jq`, we now have an additional mechanism for selective opt-out - they can remove `json.lua` from the `previewers`, that's enough.
